### PR TITLE
refactor(userspace/libsinsp)!: introducing plugin manager

### PR DESCRIPTION
**Pull request by jasondellaluce**
**What type of PR is this?**

/kind cleanup

/kind design

/kind feature

**Any specific area of the project related to this PR?**

/area libsinsp

**What this PR does / why we need it**:

This PR refactors the current libsinsp implementation of the plugin system by introducing the notion of "plugin manager". The manager is responsible of storing the plugins loaded at runtime by inspector instances and managing their state.

The manager is also used to implement optimized lookups, such as the source-compatibility check performed during field extraction (which used to be string-based and is now index-based). Right now, this should make a good performance optimization for plugin-based field extraction.

On top of that, the manager is also now the source of truth for the sources loaded at runtime, which can be useful in the future to support multiples source active at runtime.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

This PR brings few breaking changes in the `sinsp` class contract, which now exposes a new `get_plugin_manager()` method.

Moreover, the diff of the PR is big due to the git blame being moved across the splitted source files. The real diff size should be ~400 LOC.

**Does this PR introduce a user-facing change?**:

```release-note
refactor(userspace/libsinsp): introducing manager class for plugins
```

### CODE REVIEW

**Comment by FedeDP** [comment]
I really like the idea! Code is much cleaner and simple this way, imo. 
That is the standard we have been using in libsinsp, so... :)

File: userspace/libsinsp/plugin_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		return plugin_by_id(*(int32_t *)parinfo->m_val);
+	}
+	return nullptr;
+}
+
+size_t sinsp_plugin_manager::source_idx_by_plugin_id(uint32_t plugin_id, bool& found) const
+{
+	auto it = m_plugins_id_source_index.find(plugin_id);
+	found = it != m_plugins_id_source_index.end();
+	return found ? it->second : 0;
```
Comment (FedeDP) [inline userspace/libsinsp/plugin_manager.cpp:100]: Of course 0 is not a valid value there, i assume :)

**Review by FedeDP — DISMISSED** [review]
Left a minor comment; aside from that, this LGTM!
/approve

File: userspace/libsinsp/plugin_manager.cpp (hunk 1 above)
Comment (jasondellaluce) [reply userspace/libsinsp/plugin_manager.cpp:100]: The semantic here is that if `bool& found` is set to false, the returned value is not significant. If you prefer, we can declare an "invalid return value" and return that instead, making it useless to have `bool& found` in the first place. Let me know what works best for you!

Comment (FedeDP) [reply userspace/libsinsp/plugin_manager.cpp:100]: Uh didn't notice that! Sorry, it's good for me ofc!

**Review by Andreagit97 — COMMENTED** [review]
Thank you @jasondellaluce! I think that this approach could significantly decrease the complexity around plugins :rocket: I left some minor inline comments

File: userspace/libsinsp/version.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		return std::to_string(m_version_major)
+			+ "." + std::to_string(m_version_minor)
+			+ "." + std::to_string(m_version_patch);
+	}
+
+	inline bool check(const sinsp_version &requested) const
+	{
+		if(this->m_version_major != requested.m_version_major)
+		{
+			// major numbers disagree
```
[the remark this answers is not in the document]
Comment (jasondellaluce) [reply userspace/libsinsp/version.h:56]: Since this implements a standard semantic versioning check, I don't see much value in printing out which part of the version check failed. If error handling is a concern, I would much prefer having more expressive errors at the level of the callers of this function.

**Comment by jasondellaluce** [comment]
Thanks you Andrea! I addressed your comments. On top of that, I realized that the getter methods of the plugin manager could be inlined to achieve better performance. I pushed a commit with all these changes, let me know what you think.

File: userspace/libsinsp/version.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Andreagit97) [reply userspace/libsinsp/version.h:56]: Yeah maybe you are right, since it is a standard check there is no need for further info. :+1:

File: userspace/libsinsp/plugin_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+void sinsp_plugin_manager::add(std::shared_ptr<sinsp_plugin> plugin)
+{
+	for(auto& it : m_plugins)
+	{
+		// todo: we may consider dropping this constraint in the future
+		if(it->name() == plugin->name())
+		{
+			throw sinsp_exception(
+				"found multiple plugins with name " + it->name() + ". Aborting.");
```
Comment (FedeDP) [inline userspace/libsinsp/plugin_manager.cpp:37]:
```suggestion
				"found another plugin with name " + it->name() + ". Aborting.");
```

I'd rephrase it like this; i find it more correct. Wdyt?

**Review by Andreagit97 — APPROVED** [review]
/approve
